app/controllers/topological_inventory/ingress_api/v0/inventory_controller.rb

```diff
@@ -1,19 +1,31 @@
+require "topological_inventory/ingress_api"
 require "topological_inventory/ingress_api/messaging_client"
 
 module TopologicalInventory
   module IngressApi
     module V0
       class InventoryController < ApplicationController
         skip_before_action :verify_authenticity_token
+
         def save_inventory
-          payload = JSON.parse(request.body.read)
+          json_payload = request.body.read
+          payload      = JSON.parse(json_payload)
+
+          root = OpenAPIParser.parse(TopologicalInventory::IngressApi::Docs["0.0"].content,
+                                     :coerce_value => true, :datetime_coerce_class => DateTime)
+
+          request_operation = root.request_operation(:post, '/inventory')
+
+          request_operation.validate_request_body('application/json', payload)
 
           retry_count = 0
           retry_max   = 1
 
           TopologicalInventory::IngressApi.with_messaging_client do |client|
             begin
-              client.publish_message(save_inventory_payload(payload))
+              # TODO(lsmola) can use this after https://github.com/ManageIQ/manageiq-messaging/pull/45 is released
+              # client.publish_message(save_inventory_payload(json_payload))
+              client.publish_message(save_inventory_payload(JSON.parse(json_payload)))
```

Comment:
@Ladas is there a reason you're parsing this twice?  Why not just keep using the already parsed `payload`?  Does the validation modify the payload in any way?

Reply:
yes, the validation does modify it, it's using inplace coercing for ```:coerce_value => true, :datetime_coerce_class => DateTime``` and that makes the body bigger, ending with blowing up the kafka limit

Reply:
Okay that's unfortunate, can we make a copy of the payload then instead of parsing it again?

Reply:
so we need to either implement the messaging option listed ^, or extend persister so it can receive string and do the JSON parsing

Reply:
So can you .dup the payload before you validate it?

Reply:
I think the .dup will just do a shallow copy?

Reply:
https://api.rubyonrails.org/classes/Hash.html#method-i-deep_dup ?

It is a lot less intensive of an op to copy some memory than to re-parse the json object.